### DISCUSSION TITLE
Building JAX port of Detectron2

### DISCUSSION THREAD

**Discussion by rozeappletree** [google/flax#2085] [Show and tell]
Hi, I have been using Detectron2 for almost a year for training large scale models and their results (mask rcnn) always fascinate me. Like FastAI it is one of the most flexible frameworks I have used for deeplearning.

I wanted to understand internals of both JAX and Detectron2 for some time and I think building flax port for Detectron2 will get me there.

> Can I know if I can get any support from the team for solving issues if I get stuck.

Timeline:

6mos to 1year starting from Feb 2022. I have couple of commitments that hinder me from full time contribution but can spare weekends and some weekdays for the project. The idea is to slowly but effectively build the project.

Motivation:

Building mature frameworks for training models (stress on large-scale) is always critical for both research and industry. And JAX is currently the hottest topic I want to learn by doing.

**Comment by rozeappletree** [comment]
Linking discussion from Detectron2 https://github.com/facebookresearch/detectron2/discussions/3817

**Comment by avital** [comment]
Hi @INF800 -- we'd love to see something like Detectron2 ported to JAX/Flax. (Though I personally don't know almost anything about that project).

What kind of support would you be looking for? Then we can ask around and see if someone can help in the best way.

**Comment by rozeappletree** [comment]
Hi thank you for the response. I will NOT be needing any help for
understanding the already implemented code in Detectron2 (or any python
code as such).

But Instead, I will be needing help in porting pytorch code to JAX (i.e
FLAX) with best practices. In JAX you can consider me at beginner level and
for pytorch I am at intermediate level.

**Comment by avital** [comment]
Folks on our GitHub discussion board are pretty helpful, and both folks that work on JAX and Flax, as well as knowledgable community members can help. The upside this way is that other people can benefit from the learning. I propose we start with that, and if you start making progress and would benefit from more intensive support we can see what we can do. How do you feel about that?

**Comment by rozeappletree** [comment]
> Folks on our GitHub discussion board are pretty helpful, and both folks that work on JAX and Flax, as well as knowledgable community members can help. The upside this way is that other people can benefit from the learning. I propose we start with that, and if you start making progress and would benefit from more intensive support we can see what we can do. How do you feel about that?

Perfect! I will open up related issues/discussions and  link them to this issue to keep track of them. As said before I will be starting it actively from Feb 2022. Until then will try to understand more about JAX and Detectron2 and you can expect me to open up multiple discussion topics related to JAX :)

**Comment by avital** [comment]
Great! One possible first step is only to port one particular model in a standalone codebase, such as Mask R-CNN, and see what we need to do from there.

**Comment by rozeappletree** [comment]
> Great! One possible first step is only to port one particular model in a standalone codebase, such as Mask R-CNN, and see what we need to do from there.

Makes sense I was planning to do the same. Will update here once I create it.

**Comment by saberkun** [comment]
https://github.com/google-research/scenic is a flax computer vision code base as a reference.

**Comment by rozeappletree** [comment]
> https://github.com/google-research/scenic is a flax computer vision code base as a reference.

Thanks. Detectron2 internally uses fvcore https://github.com/facebookresearch/fvcore. Is this project similar to fvcore?

**Comment by marcvanzee** [comment]
@INF800 I moved your issue to a "Show and Tell" Discussion Thread, which seemed more appropriate. Looking forward to hearing about your progress!
